accepted/2024/observable_collection.md

```diff
@@ -0,0 +1,173 @@
+
+### Non-Goals
+
+* Support bulk notifications for discontinuous ranges in the collection
+
+## Stakeholders and Reviewers
+
+* Platform components
+    - Libraries team
+    - WinForms team
+    - WPF team
+    - MAUI team
+* External UI platforms
+    - Avalonia
+    - Uno
+
+## Design
+
+### API Changes
+
+We don't need to make any changes to `INotifyCollectionChanged` as
+`NotifyCollectionChangedEventArgs` can already represent ranges:
+
+```C#
+public partial class NotifyCollectionChangedEventArgs : EventArgs
+{
+    // Constructors omitted
+    public NotifyCollectionChangedAction Action { get; }
+    public IList? NewItems { get; }
+    public IList? OldItems { get; }
+    public int NewStartingIndex { get; }
+    public int OldStartingIndex { get; }
+}
+```
+
+> [!NOTE]
+> Please note that the API shape only supports batched notifications if the
+> affected range is contiguous, for example, inserting multiple items or
+> removing a contiguous range. It doesn't, for example, support removing
+> multiple discontiguous items.
+
+`ObservableCollection<T>` needs to expose new APIs for performing bulk
+operations that will result in bulk notifications:
+
+```C#
+namespace System.Collections.ObjectModel;
+
+public class ObservableCollection<T>
+{
+    public void AddRange(IEnumerable<T> collection);
+    public void InsertRange(int index, IEnumerable<T> collection);
+    public void RemoveRange(int index, int count);
+    public void ReplaceRange(int index, int count, IEnumerable<T> collection);
+
+    protected virtual void InsertItemsRange(int index, IEnumerable<T> collection);
+    protected virtual void RemoveItemsRange(int index, int count);
+    protected virtual void ReplaceItemsRange(int index, int count, IEnumerable<T> collection);
+}
+```
+
+## How it breaks consumers
+
+`ObservableCollection<T>` exists since .NET Framework 3.0, which was shipped in
+2006, 18 years ago. While the eventing structure in principle supports raising
+events for more than one item, the collection was never capable of raising such
+events. As a result, consumers of `ObservableCollection<T>`, specifically WPF,
+have baked in assumptions that these can never happen. As a result, when tried
+to make these changes a ton of code in WPF broke, i.e. it crashed with
+exceptions.
+
+## Options to avoid breaking consumers
+
+* **Do nothing**. We could just do nothing and simply let customers report bugs
+  for controls that don't support bulk notifications. Strictly speaking, this
+  isn't a behavioral change as existing code won't start raising bulk events
+  unless the user is calling the new methods on `ObservableCollection<T>`. The
+  guidance would be "use bulk operations if the data bound UI supports it,
+  otherwise don't". We would treat the lack of WPF support as a feature request
+  and let it run its course.
+  - Pro: Simplest API, easiest to implement, unblocks the ecosystem.
+  - Con: Violates our goal to not ship functionality we know won't work with
```

Comment:
Other changes have been made between .NET versions which have required work on WPF and other teams (like WinUI), for instance RID change. So, doesn't this just involve internal planning/coordination of the normal engineering work involved with updating .NET versions? This has been the sticking point, so it'd just be a forcing function otherwise, right?

tldr - this is what we've been trying to do and it hasn't worked up until now... so we should either just do it and force the debt upon others early or evaluate the other options, right?

Reply:
I'm not even sure that the change in WPF would be significant -- once we get a dev to do thorough investigation on what fails and identifies what changes are actually necessary. So far, the point in times where the issue was discovered simply didn't give us enough reaction time to do that.

> Doesn't this just involve internal planning/coordination of the normal engineering work involved with updating .NET versions?

I believe that is true. But of course, this assumes that the teams having to do work buy into the value prop. RID changes were really driven by changes necessary at the bottom. This is different here because it's really feature work we're asking for.